File: solutions/python/rest-api/1/rest_api.py
import json

class RestAPI:
    def __init__(self, database=None):
        self.users = {}
        if database and "users" in database:
            for user in database["users"]:
                name = user["name"]
                self.users[name] = {"owes": user["owes"].copy(), "owed_by": user["owed_by"].copy()}

    def get(self, url, payload=None):
        if payload:
            payload = json.loads(payload)
        if url == "/users":
            users_data = self.users
            if payload and "users" in payload:
                requested_users = payload["users"]
                users_data = {name: self.users[name] for name in requested_users if name in self.users}
            
            result = {"users": []}
            for name in sorted(users_data):
                user = users_data[name]
                owes_total = sum(user.get("owes", {}).values())
                owed_by_total = sum(user.get("owed_by", {}).values())
                balance = owed_by_total - owes_total
                user_data = {
                    "name": name,
                    "owes": user.get("owes", {}),
                    "owed_by": user.get("owed_by", {}),
                    "balance": balance
                }
                result["users"].append(user_data)
            return json.dumps(result)
        return json.dumps({"error": "Invalid URL"})

    def post(self, url, payload=None):
        if payload:
            payload = json.loads(payload)
        if url == "/add" and payload and "user" in payload:
            new_user = payload["user"]
            if new_user not in self.users:
                self.users[new_user] = {"owes": {}, "owed_by": {}}
                owes_total = sum(self.users[new_user].get("owes", {}).values())
                owed_by_total = sum(self.users[new_user].get("owed_by", {}).values())
                balance = owed_by_total - owes_total
                user_data = {
                    "name": new_user,
                    "owes": self.users[new_user]["owes"],
                    "owed_by": self.users[new_user]["owed_by"],
                    "balance": balance
                }
                return json.dumps(user_data)
            return json.dumps({"error": "User already exists"})

        elif url == "/iou" and payload and all(key in payload for key in ["lender", "borrower", "amount"]):
            lender = payload["lender"]
            borrower = payload["borrower"]
            amount = payload["amount"]

            if lender == borrower:
                return json.dumps({"error": "Lender and borrower cannot be the same"})

            if lender not in self.users:
                self.users[lender] = {"owes": {}, "owed_by": {}}
            if borrower not in self.users:
                self.users[borrower] = {"owes": {}, "owed_by": {}}

            # Calculate net
            existing_borrower_owes_lender = self.users[borrower]["owes"].get(lender, 0.0)
            existing_lender_owes_borrower = self.users[lender]["owes"].get(borrower, 0.0)
            net_existing = existing_borrower_owes_lender - existing_lender_owes_borrower
            new_net = net_existing + amount

            # Remove existing entries
            self.users[borrower]["owes"].pop(lender, None)
            self.users[lender]["owed_by"].pop(borrower, None)
            self.users[lender]["owes"].pop(borrower, None)
            self.users[borrower]["owed_by"].pop(lender, None)

            if new_net > 0:
                self.users[borrower]["owes"][lender] = new_net
                self.users[lender]["owed_by"][borrower] = new_net
            elif new_net < 0:
                self.users[lender]["owes"][borrower] = -new_net
                self.users[borrower]["owed_by"][lender] = -new_net
            # else: 0, do nothing

            result = {"users": []}
            for name in sorted([lender, borrower]):
                user = self.users[name]
                owes_total = sum(user.get("owes", {}).values())
                owed_by_total = sum(user.get("owed_by", {}).values())
                balance = owed_by_total - owes_total
                user_data = {
                    "name": name,
                    "owes": user["owes"],
                    "owed_by": user["owed_by"],
                    "balance": balance
                }
                result["users"].append(user_data)
            return json.dumps(result)
        return json.dumps({"error": "Invalid payload or URL"})

File: solutions/javascript/rest-api/1/rest-api.js
export class RestAPI {
  constructor(db = { users: [] }) {
    // Deep clone the database to avoid mutating the original reference
    this.db = JSON.parse(JSON.stringify(db));
  }

  get(url) {
    if (url === '/users') {
      return { 
        users: this.db.users.sort((a, b) => a.name.localeCompare(b.name)) 
      };
    }

    if (url.startsWith('/users?users=')) {
      // Extract the comma-separated names from the query string
      const query = url.split('=')[1];
      const names = query.split(',');
      
      const filteredUsers = this.db.users
        .filter(u => names.includes(u.name))
        .sort((a, b) => a.name.localeCompare(b.name));
        
      return { users: filteredUsers };
    }

    throw new Error('Endpoint not found');
  }

  post(url, payload) {
    if (url === '/add') {
      const newUser = {
        name: payload.user,
        owes: {},
        owed_by: {},
        balance: 0.0
      };
      
      this.db.users.push(newUser);
      return newUser;
    }

    if (url === '/iou') {
      const { lender, borrower, amount } = payload;
      
      const lenderObj = this.db.users.find(u => u.name === lender);
      const borrowerObj = this.db.users.find(u => u.name === borrower);

      // 1. Calculate the net balance between these two specific users
      const currentOwedByBorrower = lenderObj.owed_by[borrower] || 0;
      const currentOwesBorrower = lenderObj.owes[borrower] || 0;
      const net = currentOwedByBorrower - currentOwesBorrower + amount;

      // 2. Update Lender's records based on the new net
      if (net > 0) {
        lenderObj.owed_by[borrower] = net;
        delete lenderObj.owes[borrower];
      } else if (net < 0) {
        lenderObj.owes[borrower] = -net;
        delete lenderObj.owed_by[borrower];
      } else {
        delete lenderObj.owed_by[borrower];
        delete lenderObj.owes[borrower];
      }

      // 3. Update Borrower's records (the mathematical inverse of the lender)
      const borrowerNet = -net;
      if (borrowerNet > 0) {
        borrowerObj.owed_by[lender] = borrowerNet;
        delete borrowerObj.owes[lender];
      } else if (borrowerNet < 0) {
        borrowerObj.owes[lender] = -borrowerNet;
        delete borrowerObj.owed_by[lender];
      } else {
        delete borrowerObj.owed_by[lender];
        delete borrowerObj.owes[lender];
      }

      // 4. Helper function to recalculate a user's total balance
      const updateBalance = (user) => {
        const totalOwedBy = Object.values(user.owed_by).reduce((acc, val) => acc + val, 0);
        const totalOwes = Object.values(user.owes).reduce((acc, val) => acc + val, 0);
        user.balance = totalOwedBy - totalOwes;
      };

      updateBalance(lenderObj);
      updateBalance(borrowerObj);

      // Return both updated users, sorted alphabetically by name
      return {
        users: [lenderObj, borrowerObj].sort((a, b) => a.name.localeCompare(b.name))
      };
    }

    throw new Error('Endpoint not found');
  }
}
